### ISSUE TITLE
Make the crawler concurrent

### ISSUE THREAD

**Issue by rajivharlalka** [metakgp/iqps-go#91]
**Is your feature request related to a problem? Please describe.**
Currently the crawler sequentially fetches each paper details, parses it and downloads the paper. This can be made lot faster using go-routines.

**Comment by rajivharlalka** [comment]
Did the implementation have an upperbound on the number of parallel requests being made? AFAIR no. IMO using waitgroups to limit the number of concurrent workers to 2-3 should improve the perform significantly.

**Comment by shikharish** [comment]
I dont remember exactly.
BTW we won't need to implement go-routines ourselves as colly has a option to enable async request and also limit them. Can test that.

**Comment by proffapt** [comment]
@rajivharlalka or @harshkhandeparkar please update the state of this issue to be reflected on the kanban.

**Comment by harshkhandeparkar** [comment]
@shikharish what should be the status of this?

**Comment by shikharish** [comment]
It is not needed as of now. We only need to run the crawler once or twice a semester so it's very low priority.

**Comment by harshkhandeparkar** [comment]
Is it hard to do?

**Comment by harshkhandeparkar** [comment]
Then just finish it off maybe?

**Comment by harshkhandeparkar** [comment]
No point in keeping hanging issues if they can be solved in a few minutes.

**Comment by shikharish** [comment]
Ok please assign it to me

**Comment by proffapt** [comment]
@shikharish updates?

**Comment by shikharish** [comment]
Did some testing and turns out even using 2 go routines leads to dropping of 1-2 requests. Further increasing it to 6 goroutines makes it 3-4 requests.
 
Should we skip this one for now?

**Comment by proffapt** [comment]
Try to implement a retry function. Also, how many requests are you able to make concurrently. Even if it is more than one, then that's a win situation.

**Comment by proffapt** [comment]
Halting this, till we have time to have a look at it comfortably.

**Comment by harshkhandeparkar** [comment]
@dipamsen does your PR handle this?

**Comment by dipamsen** [comment]
It does not, I could try doing it

**Comment by harshkhandeparkar** [comment]
> It does not, I could try doing it

How fast is it in the current state?

**Comment by dipamsen** [comment]
Based on a rough measurement ~ 1-2 min for all papers in 2024

**Comment by harshkhandeparkar** [comment]
> Based on a rough measurement ~ 1-2 min for all papers in 2024

Hmm, see if concurrency can help. See the above messages as well, and make it a CLI option if possible.

**Comment by dipamsen** [comment]
Just to clarify, I have to make the downloading of pdfs concurrent, right? @harshkhandeparkar

**Comment by harshkhandeparkar** [comment]
> Just to clarify, I have to make the downloading of pdfs concurrent, right? [@harshkhandeparkar](https://github.com/harshkhandeparkar)

Yes. However, parsing can also be concurrent, but download would be the limiting, I presume, unless OCR is done.
